contrib/mesos/pkg/minion/tasks/task.go

```diff
@@ -0,0 +1,369 @@
+	writer := t.createLogger()
+	go func() {
+		defer writer.Close()
+		_, err := io.Copy(writer, r)
+		if err != nil && err != io.EOF {
+			// using tryComplete is racy because the state machine closes completedCh and
+			// so we don't want to attempt to write to a closed/closing chan. so
+			// just log this for now.
+			log.Errorf("logger for task %q crashed: %v", t.bin, err)
+		}
+	}()
+}
+
+// notStartedTask spawns the given task and transitions to a startedTask state
+func notStartedTask(t *Task) taskStateFn {
+	log.Infof("starting task process %q with args '%+v'", t.bin, t.args)
+
+	// create command
+	cmd := exec.Command(t.bin, t.args...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		t.tryError(fmt.Errorf("error getting stdout of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		t.tryError(fmt.Errorf("error getting stderr of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+
+	t.initLogging(stderrLogs)
+	if len(t.env) > 0 {
+		cmd.Env = t.env
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid:   true,
+		Pdeathsig: syscall.SIGKILL, // see cmdProcess.Kill
+	}
+
+	// last min check for shouldQuit here
+	select {
+	case <-t.shouldQuit:
+		t.tryError(fmt.Errorf("task execution canceled, aborting process launch"))
+		return taskShouldRestart
+	default:
+	}
+
+	if err := cmd.Start(); err != nil {
+		t.tryError(fmt.Errorf("failed to start task process %q: %v", t.bin, err))
+		return taskShouldRestart
+	}
+	log.Infoln("task started", t.name)
+	t.cmd = &cmdProcess{delegate: cmd}
+	return taskRunning
+}
+
+type exitError interface {
+	error
+
+	// see os.ProcessState.Sys: returned value can be converted to something like syscall.WaitStatus
+	Sys() interface{}
+}
+
+func taskRunning(t *Task) taskStateFn {
+	waiter := t.cmd.Wait
+	type waitResult struct {
+		completed *TaskCompletion
+		err       error
+	}
+	var sendOnce sync.Once
+	trySend := func(wr *waitResult) {
+		// guarded with once because we're only allowed to send a single "result" for each
+		// process termination. for example, if Kill() results in an error because Wait()
+		// already completed we only want to return a single result for the process.
+		sendOnce.Do(func() {
+			if wr.err != nil {
+				t.tryError(wr.err)
+			} else if wr.completed != nil {
+				t.tryComplete(wr.completed)
+			}
+		})
+	}
+	// listen for normal process completion in a goroutine; don't block because we need to listen for shouldQuit
+	waitCh := make(chan *waitResult, 1)
+	go func() {
+		wr := &waitResult{}
+		defer func() {
+			waitCh <- wr
+			close(waitCh)
+		}()
+
+		if err := waiter(); err != nil {
+			if exitError, ok := err.(exitError); ok {
+				if waitStatus, ok := exitError.Sys().(syscall.WaitStatus); ok {
+					wr.completed = &TaskCompletion{
+						name: t.name,
+						code: waitStatus.ExitStatus(),
+					}
+					return
+				}
+			}
+			wr.err = fmt.Errorf("task wait ended strangely for %q: %v", t.bin, err)
```

Comment:
Simpler: create a TaskCompletion event here with the err. Then get rid of the extra waitResult struct type above.